feat: add mkImportJava for CodeActionProvider

stormckey opened the pull request:
According to #9125, we add `mkImportJava` to automatically import a undefined name/class if it's an existing class on the Java platform.

Note that a new object ClassMap is introduced to quickly lookup the path for a  given name. Currently `ClassMap` is built by processing the string list of java files from `ClassList`.

stormckey commented:
For undefined class:
<img width="315" alt="image" src="https://github.com/user-attachments/assets/7aa232a6-81c6-4565-bbf2-a0a88a145b5e">
<img width="334" alt="image" src="https://github.com/user-attachments/assets/5cf6503d-a279-46bc-9c08-30a34ebae915">

For undefined name:
<img width="233" alt="image" src="https://github.com/user-attachments/assets/5d8b7b12-ea60-4c91-8bc0-5e55d663eb6d">
<img width="380" alt="image" src="https://github.com/user-attachments/assets/010471cd-86a4-4d17-b68f-1a6f4a210299">

Currently we will not offer completion for cases like `Math.sin` since it is treated as a whole name by the error, which means we will need to process the string to find possible imports.

magnus-madsen commented:
Looks great. I added some comments on how to improve the code.

magnus-madsen commented:
Conflicted.

magnus-madsen commented:
I ran out of time today, so I will have to give more feedback tomorrow.

However, one thing you can already look into is inserting the import in the right place.

Here is what I mean:

```
mod Abc {
  mod Def {
    def f(): FileReader
  }
}
```

then the import should go here:

```
mod Abc {
  mod Def {
    import java......FileReader
    def f(): FileReader
  }
}
```

To help with this, I have added `EnclosingMod` to UndefinedName. See here:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/errors/ResolutionError.scala#L672

This tells you the position of "mod Def" in the above example. You can then insert the import on the next line (properly indented).

stormckey commented:
I've adjusted the place to insert the import, preview:
<img width="331" alt="image" src="https://github.com/user-attachments/assets/7ead0090-a998-4f7a-9aea-ff0b43bb1f53">
<img width="356" alt="image" src="https://github.com/user-attachments/assets/f0adbdc8-dce9-4186-a534-04011903223d">

Note:
- I've added yet another `sourcePosition2Position` here, I think it's the third time.
- I've commented out the `mkImportJava` under `UndefinedJvmClass`, where the enclosing module is not available. Thus cases like `new File("data.txt")` is no longer supported. We need to first update `UndefinedJvmClass` to support this.

stormckey commented:
Known bug:

where there is an enclosing module from the start of the file, the position provided by `em` is  (1, 1). So we should add the import in the second line:
<img width="306" alt="image" src="https://github.com/user-attachments/assets/1342ddd3-86f0-46de-ab2d-1c306cb74994">

But when there are no enclosing module at all , the position provided by `em` is still (1, 1). However, this time we can't always insert in the second line:
<img width="221" alt="image" src="https://github.com/user-attachments/assets/011d43c1-55c2-4cb4-8fe8-d4c022bfdcd5">
<img width="305" alt="image" src="https://github.com/user-attachments/assets/d2d9fda1-851b-4918-b816-b00569c93ef7">

I think we need a mechanism to distinguish them.

magnus-madsen commented:
> I've commented out the mkImportJava under UndefinedJvmClass, where the enclosing module is not available. Thus cases like new File("data.txt") is no longer supported. We need to first update UndefinedJvmClass to support this.

What is this about?

magnus-madsen commented:
For the known bug, perhaps I should give the source location of the entire module? Plus some information of whether it is root.

magnus-madsen commented:
We have added: https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/ast/shared/AnchorPosition.scala for your use.

magnus-madsen commented:
An UndefinedName/UndefinedType now has an anchor position and you can use that directly. It should have the "right" line number.

magnus-madsen commented:
> I've commented out the mkImportJava under UndefinedJvmClass, where the enclosing module is not available. Thus cases like new File("data.txt") is no longer supported. We need to first update UndefinedJvmClass to support this.

I understand your comment now. 

I think a very small extension here:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Resolver.scala#L1333

similar to 

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/Resolver.scala#L2262

Will get you exactly what you need.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/CodeActionProvider.scala:
```diff
@@ -218,6 +222,44 @@ object CodeActionProvider {
+    *   import java.io.File
+    * }}}
+    */
+  private def mkImportJava(name: String, uri: String, em: EnclosingMod): List[CodeAction] = {
```
I renamed EnclosingMod to AnchorPosition and changed its definition, so you will need to merge/rebase and then refactor.

stormckey commented:
Now `mkImportJava` uses `AnchorPosition` :) . But there is still something to note:

When inserting in `(line+1, col)`, there are several problems:

No one could assure that there is nothing in this place(but it's abnormal), for example:
<img width="289" alt="image" src="https://github.com/user-attachments/assets/c570e294-df94-40a4-b2e5-33e1a370cb9d">
Then the import will break the comment:
<img width="339" alt="image" src="https://github.com/user-attachments/assets/f3b44d4a-e9b5-410d-9bc5-3e8d043ea562">

Also, we need to manually add spaces after the import to make the indentation of next line works fine(because our import takes part of its indentation).

I used to insert in the beginning of the next line and add the spaces for import manually, which do you think is better?

magnus-madsen commented:
Is it possible to ask to insert and entire new line? Or maybe we need to insert "\n<spaces>import foo" ?

magnus-madsen commented:
(There is also a conflict for you to fix)

stormckey commented:
> Is it possible to ask to insert and entire new line? Or maybe we need to insert "\nimport foo" ?

If we want to avoid the first problem totally, we'll need to insert in the beginning the next line, so the text after insertion will be"import xxx\n//This is B!". Otherwise it will still be "//This\nimport xxx\nis B!"

stormckey commented:
> (There is also a conflict for you to fix)

It keeps to think of `mkImportJava` and `mkNewDef` as a conflict, which is weird. Why not just keep both🤔.

magnus-madsen commented:
(We can meet on Zoom now, if you have time)

stormckey commented:
> (We can meet on Zoom now, if you have time)

let'e meet.

magnus-madsen commented:
See https://github.com/flix/flix/pull/9222

stormckey commented:
My bad, this should be sp.col + 3 (short for - 1 + 4) to make the number of spaces correct.
https://github.com/flix/flix/blob/97b568c87d30da6bf1bb1fe3947c570d5e5fab18/main/src/ca/uwaterloo/flix/language/ast/shared/AnchorPosition.scala#L37-L43

Shall I change the code here in this pr?

magnus-madsen commented:
Yes, please make the change. Please write is as +4 -1

magnus-madsen commented:
Also post some screenshots when you have it working :)

stormckey commented:
Here is the preview:
<img width="351" alt="image" src="https://github.com/user-attachments/assets/522fe463-d832-4374-9855-fecd93f1fca6">
<img width="327" alt="image" src="https://github.com/user-attachments/assets/d369327e-037e-4826-9390-0e8bff06a99f">

on Root:
<img width="305" alt="image" src="https://github.com/user-attachments/assets/e4722585-33ef-482a-9281-4bbf75a3c03f">
<img width="282" alt="image" src="https://github.com/user-attachments/assets/e99bef2d-0c57-42ca-b5fd-19c391b70ec3">

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/CodeActionProvider.scala:
```diff
@@ -43,8 +44,11 @@ object CodeActionProvider {
 
-    case ResolutionError.UndefinedName(qn, _, env, _, loc) if overlaps(range, loc) =>
-      mkNewDef(qn.ident.name, uri) :: {
+//    case ResolutionError.UndefinedJvmClass(name, _, loc) if onSameLine(range, loc) =>
```
Is this for a followup PR?

stormckey replied:
Yeah, we'll need to first extend UndefinedJvmClass and then complete the code, as discussed above .

magnus-madsen replied:
We do it after this has been reviewed and merged 👍

magnus-madsen commented on main/src/ca/uwaterloo/flix/util/ClassList.scala:
```diff
@@ -15,22 +15,37 @@
+   * The map from class name to class path, built from TheList
+   *
+   * Example:
+   *  - List -> java.util.List
```
The right should be a set, e.g. List -> {java.util.List} to highlight the multimappedness

magnus-madsen commented:
Thanks, I will take a more detailed look later.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/CodeActionProvider.scala:
```diff
@@ -249,6 +253,39 @@ object CodeActionProvider {
+    val startPosition = Position(line = ap.line, character = ap.col)
+    val insertRange = Range(startPosition, startPosition)
+    val leadingSpaces = " " * ap.spaces
+    ClassList.TheMap.get(name).toList.flatten.map { path =>
```
toList.flatten.map -- I think this can be simplified? If nothing else, with just an ordinary pattern match

stormckey replied:
The type after get is Option[Set[String]]. If using pattern match , we'll need first to process the option then map the Set to List, which I think is more complicated than this.🤔 Or you have other suggestions?

magnus-madsen commented:
I added one comment. But I will merge now, so you can fix it later.
